### PR TITLE
CORS Access Error Fix

### PR DESCRIPTION
**Pull request by 98farhan94**
### Proposed changes in this pull request
Access control headers added to response to fix the preflight test issue because of CORS leading to network error over TLS query.

screen shot 2018-09-12 at 9 41 20 am

Actual Error: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

### Type (put an `x` where ever applicable)
- [x ] Bug fix: Link to the issue
- [ ] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [x] Read the [CONTRIBUTION guidelines](https://github.com/nknorg/nkn#contributing).
- [ ] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [ ] Added proper documentation where ever applicable (in code and README.md).
- [x] Code has been written according to [NKN-Golang-Style-Guide](https://github.com/nknorg/nkn/wiki/NKN-Golang-Style-Guide)

### Extra information
RPC calls were failing over TLS because of missing access headers.

### CODE REVIEW

**Comment by yilunzhang** [comment]
Should we make CORS optional in case some people don't want their node to be access cross site?

**Comment by 98farhan94** [comment]
@yilunzhang You are right, it can become a potential attack vector. It will better for security to have it disabled by default.

In my case I am using the `<domain>` instead of `*` (allow all) to make sure only my website can access it.

**Comment by doraemon0** [comment]
The CORS code already existed:
https://github.com/nknorg/nkn/blob/master/api/httpjson/RPCserver.go#L57

It is being called at line:
https://github.com/nknorg/nkn/blob/master/api/httpjson/RPCserver.go#L133
https://github.com/nknorg/nkn/blob/master/api/httpjson/RPCserver.go#L150

Please help to check the root cause.

**Comment by 98farhan94** [comment]
@doraemon0 I did notice the presence of CORS in that line, it just wasn't being called. It seems CORS "preflight" request is an `OPTIONS` request instead of `POST` ([Mozilla Developer](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request)) .

Could it be that because it's an OPTIONS request it gets deferred and does not use s.write function, and goes to the default request instead?
https://github.com/nknorg/nkn/blob/master/api/httpjson/RPCserver.go#L57

Am I right to think this is the function that gets called on preflight check? I could be wrong, I haven't worked with RPC before.
https://github.com/nknorg/nkn/blob/master/api/httpjson/RPCserver.go#L162

**Comment by yilunzhang** [comment]
It's not ideal to have both places set CORS, and from the code they should both work or not work. Just to make sure: could you please do 2 tests, one with CORS only in write (without your patch), and one with only CORS only in handle (remove the existing ones in write)? If only one works, we should remove the other.

File: api/httpjson/RPCserver.go [hunk 1]
```diff
@@ -64,6 +64,9 @@ func (s *RPCServer) write(w http.ResponseWriter, data []byte) {
 //this is the funciton that should be called in order to answer an rpc call
 //should be registered like "http.HandleFunc("/", httpjsonrpc.Handle)"
 func (s *RPCServer) Handle(w http.ResponseWriter, r *http.Request) {
+	w.Header().Add("Access-Control-Allow-Headers", "Content-Type")
+	w.Header().Set("content-type", "application/json;charset=utf-8")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
```
Comment (doraemon0) [inline api/httpjson/RPCserver.go]: I prefer move these lines to the read lock.

**Comment by 98farhan94** [comment]
Without the fix:
screen shot 2018-09-13 at 2 11 01 pm

With the fix and write code removed:
screen shot 2018-09-13 at 2 14 45 pm

So I guess we can just remove the original code.

File: api/httpjson/RPCserver.go (hunk 1 above)
Comment (98farhan94) [reply api/httpjson/RPCserver.go]: Done.

File: api/httpjson/RPCserver.go [hunk 2]
```diff
@@ -55,9 +55,6 @@ func NewServer(node protocol.Noder, wallet vault.Wallet) *RPCServer {
 }
 
 func (s *RPCServer) write(w http.ResponseWriter, data []byte) {
```
Comment (doraemon0) [inline api/httpjson/RPCserver.go]: This function is no longer required. Please remove this function and change `s.write(w, data)` to `w.Write(data)` is better.

File: api/httpjson/RPCserver.go (hunk 1 above)
Comment (98farhan94) [reply api/httpjson/RPCserver.go]: Updated and also removed unused function.

File: api/httpjson/RPCserver.go (hunk 2 above)
Comment (98farhan94) [reply api/httpjson/RPCserver.go]: Did not realise, done!

**Pull request merged**